### PR TITLE
http: handle possible range requests properly.

### LINKED ISSUE #2098: api: getObjectHandler returns InvalidRange for offsets with in the length of data.
## Expected behaviour

Should not return an error for offset value panning with in the length of the data.
## Actual behaviour

We return InvalidRange error 
## Steps to reproduce the behavior

Following code runs fine on S3 but Minio server returns InvalidRange.

```
func (s *TestSuiteCommon) TestGetObjectRange(c *C) {
        // generate a random bucket name.
        bucketName := getRandomBucketName()
        // HTTP request to create the bucket.
        request, err := newTestRequest("PUT", getMakeBucketURL(s.endPoint, bucketName),
                0, nil, s.accessKey, s.secretKey)
        c.Assert(err, IsNil)

        client := http.Client{}
        // execute the HTTP request to create bucket.
        response, err := client.Do(request)
        c.Assert(err, IsNil)
        c.Assert(response.StatusCode, Equals, http.StatusOK)

        // content for the object to be uploaded.
        buffer1 := bytes.NewReader([]byte("Hello World"))

        objectName := "test-object"
        // HTTP request to upload the object.
        request, err = newTestRequest("PUT", getPutObjectURL(s.endPoint, bucketName, objectName),
                int64(buffer1.Len()), buffer1, s.accessKey, s.secretKey)
        c.Assert(err, IsNil)

        client = http.Client{}
        // execute the HTTP request to upload the object.
        response, err = client.Do(request)
        c.Assert(err, IsNil)
        // verify whether upload was successful.
        c.Assert(response.StatusCode, Equals, http.StatusOK)

        // HTTP request to download the object.
        request, err = newTestRequest("GET", getGetObjectURL(s.endPoint, bucketName, objectName),
                0, nil, s.accessKey, s.secretKey)
        // Invalid byte range set.
        request.Header.Add("Range", "bytes=7-6")
        c.Assert(err, IsNil)

        client = http.Client{}
        // execute the HTTP request.
        response, err = client.Do(request)
        c.Assert(err, IsNil)
        c.Assert(response.StatusCode, Equals, http.StatusOK)
}
```
## Minio version
- (paste output of `minio version`)
  0540863663b9f515a0a5bd803af34adca08468dc
## System information

Ubuntu 16.04

### LINKED ISSUE #2118: FS: GetObject: Incorrect behaviour with invalid offset.
## Expected behaviour

To Fail with `InvalidRange`. XL behaves as expected for this case. 
## Actual behaviour

Test passing. 
## Steps to reproduce the behaviour

Run the test here https://github.com/hackintoshrao/minio/blob/65a80044e4b6e551d903c1e91c9ad2d659c9e475/object-api-getobject_test.go 
## Minio version

Latest master at this point in time 169c72cdab14ac0f993161f6ed97e8462eccfb97 
## System information

OSX ElCaptain.

### PR DESCRIPTION
**Pull request by balamurugana**
Previously range string is not validated against various combination
of values.  This patch fixes the issue.

Fixes #2098

### CODE REVIEW

File: httprange_test.go [hunk 1]
```diff
@@ -52,16 +55,36 @@ func TestParseRequestRange(t *testing.T) {
 	}
 
 	// Test invalid range strings.
-	invalidRangeStrings := []string{"", "2-5", "bytes= 2-5", "bytes=2 - 5", "bytes=0-0,-1", "bytes=5-2", "bytes=2-5 ", "bytes=2--5"}
+	invalidRangeStrings := []string{
+		"bytes=8",
```
Comment (harshavardhana) [inline httprange_test.go:59]: why additional space?

File: httprange_test.go [hunk 2]
```diff
@@ -52,16 +55,36 @@ func TestParseRequestRange(t *testing.T) {
...
+		"2-5",
+		"bytes = 2-5",
+		"bytes=2 - 5",
+		"bytes=0-0,-1",
+		"bytes=2-5 ",
+	}
 	for _, rangeString := range invalidRangeStrings {
 		if _, err := parseRequestRange(rangeString, 10); err == nil {
 			t.Fatalf("expected: an error, got: <nil>")
 		}
```
Comment (hackintoshrao) [inline httprange_test.go]: Can the error message also be asserted.

File: httprange_test.go [hunk 3]
```diff
@@ -52,16 +55,36 @@ func TestParseRequestRange(t *testing.T) {
...
+		"bytes=-",
+		"",
+		"2-5",
+		"bytes = 2-5",
+		"bytes=2 - 5",
+		"bytes=0-0,-1",
+		"bytes=2-5 ",
+	}
 	for _, rangeString := range invalidRangeStrings {
 		if _, err := parseRequestRange(rangeString, 10); err == nil {
```
Comment (hackintoshrao) [inline httprange_test.go:73]: The error message should also be asserted here.

File: httprange_test.go [hunk 4]
```diff
@@ -52,16 +55,36 @@ func TestParseRequestRange(t *testing.T) {
...
 			t.Fatalf("expected: an error, got: <nil>")
 		}
 	}
 
 	// Test error range strings.
-	errorRangeString := "bytes=-0"
-	if _, err := parseRequestRange(errorRangeString, 10); err != errInvalidRange {
-		t.Fatalf("expected: %s, got: %s", errInvalidRange, err)
+	errorRangeString := []string{
```
Comment (hackintoshrao) [inline httprange_test.go:79]: The error cases and the succees should be formed into a single array. They should be together, the success and error cases shouldn't be separated.

**Comment by harshavardhana** [comment]
Will merge this and we can cleanup the tests subsequently.

**Pull request merged**